syntax = "proto3";

option go_package = ".;pb";

import "google/protobuf/empty.proto";
import "domain.proto";
import "volume.proto";
import "port_forwarding.proto";
import "dns.proto";
import "route.proto";

message SyncRoutesRequest {}

service ControllerService {
  rpc GetDNSRecord(DNSRecordIdentifier) returns (DNSRecord) {}
  rpc ListDNSRecords(ListDNSRecordsRequest) returns (ListDNSRecordsResponse) {}
  rpc PutDNSRecord(PutDNSRecordRequest) returns (DNSRecord) {}
  rpc DeleteDNSRecord(DNSRecordIdentifier) returns (google.protobuf.Empty) {}

  rpc GetNetwork(GetNetworkRequest) returns (Network) {}
  rpc ListNetworks(ListNetworksRequest) returns (ListNetworksResponse) {}
  rpc CreateNetwork(CreateNetworkRequest) returns (Network) {}
  rpc DeleteNetwork(DeleteNetworkRequest) returns (google.protobuf.Empty) {}

  rpc StartDomain(StartDomainRequest) returns (google.protobuf.Empty) {}
  rpc StopDomain(StopDomainRequest) returns (google.protobuf.Empty) {}
  rpc GetDomain(GetDomainRequest) returns (Domain) {}
  rpc ListDomains(ListDomainsRequest) returns (ListDomainsResponse) {}
  rpc CreateDomain(CreateDomainRequest) returns (Domain) {}
  rpc DeleteDomain(DeleteDomainRequest) returns (google.protobuf.Empty) {}

  rpc DownloadImage(DownloadImageRequest) returns (stream ImageChunk) {}

  rpc GetVolume(GetVolumeRequest) returns (Volume) {}
  rpc ListVolumes(ListVolumesRequest) returns (ListVolumesResponse) {}
  rpc CreateVolume(CreateVolumeRequest) returns (Volume) {}
  rpc UpdateVolume(UpdateVolumeRequest) returns (Volume) {}
  rpc DeleteVolume(DeleteVolumeRequest) returns (google.protobuf.Empty) {}

  rpc ListVolumeAttachments(ListVolumeAttachmentsRequest) returns (ListVolumeAttachmentsResponse) {}
  rpc GetVolumeAttachment(VolumeAttachmentIdentifier) returns (VolumeAttachment) {}
  rpc AttachVolume(VolumeAttachmentIdentifier) returns (VolumeAttachment) {}
  rpc DetachVolume(VolumeAttachmentIdentifier) returns (google.protobuf.Empty) {}

  rpc GetPortForwarding(PortForwardingIdentifier) returns (PortForwarding) {}
  rpc ListPortForwardings(ListPortForwardingsRequest) returns (ListPortForwardingsResponse) {}
  rpc PutPortForwarding(PutPortForwardingRequest) returns (PortForwarding) {}
  rpc DeletePortForwarding(PortForwardingIdentifier) returns (google.protobuf.Empty) {}

  rpc GetRouteTable(RouteTableIdentifier) returns (RouteTable) {}
  rpc ListRouteTables(ListRouteTablesRequest) returns (ListRouteTablesResponse) {}
  rpc CreateRouteTable(CreateRouteTableRequest) returns (RouteTable) {}
  rpc DeleteRouteTable(RouteTableIdentifier) returns (google.protobuf.Empty) {}

  rpc GetRoute(RouteIdentifier) returns (Route) {}
  rpc ListRoutes(ListRoutesRequest) returns (ListRoutesResponse) {}
  rpc PutRoute(PutRouteRequest) returns (Route) {}
  rpc DeleteRoute(RouteIdentifier) returns (google.protobuf.Empty) {}

  rpc SyncRoutes(SyncRoutesRequest) returns (google.protobuf.Empty) {}
}
